packaging/create_rpm

```diff
@@ -0,0 +1,665 @@
+        message = 'Tried to execute: {command}'.format(
+            # This is not necessarily exactly what we tried to execute, but is
+            # often good enough to provide pointers for troubleshooting.
+            command=' '.join(command),
+        )
+        print_in_color(message, 'red', stream=sys.stderr)
+        exit_with_sadness(error)
+
+
+def get_package_urls_file(edition, branch, repo_destination_path):
+    """
+        Get the package urls list, returning the paths to the file (in a dict)
+        for 'agent' and 'manager' package urls.
+    """
+    edition_details = EDITIONS[edition]
+    repo_path = edition_details['repo']
+    auth_required = edition_details['requires_auth']
+
+    using_credentials = False
+    print_status_update('Getting package urls.')
+    if auth_required:
+        print('Authentication is required...')
+        if (
+            'GITHUB_USERNAME' in os.environ
+            and 'GITHUB_PASSWORD' in os.environ
+        ):
+            print('Username and password have been supplied.')
+            repo = 'https://{username}:{password}@github.com/'.format(
+                username=os.environ['GITHUB_USERNAME'],
+                password=os.environ['GITHUB_PASSWORD'],
+            )
+            repo += '{repo_path}'
+            using_credentials = True
+        else:
+            print('Username and password not supplied, SSH key will be used.')
+            repo = 'git@github.com:{repo_path}'
+    else:
+        print('Authentication not required.')
+        repo = 'https://github.com/{repo_path}'
+
+    repo = repo.format(repo_path=repo_path)
+
+    if using_credentials:
+        # If credentials are used, don't spit them onto the screen.
+        # Note that if there's an error they will be dumped to the screen.
+        output_repo = repo.split('@')[-1]
+        output_repo = 'http://****:****@' + output_repo
+    else:
+        output_repo = repo
+    print('Cloning {branch} from: {repo}'.format(
+        repo=output_repo,
+        branch=branch,
+    ))
+
+    # We can't use the normal utility function for this or we will spit
+    # passwords onto the screen.
+    try:
+        subprocess.check_call([
+            'git', 'clone',
+            '--branch=' + branch,
+            '--depth=1',
+            repo,
+            repo_destination_path,
+        ])
+    except subprocess.CalledProcessError:
+        # Don't spit out the password, please.
+        exit_with_sadness(
+            'Could not git clone: {repo}'.format(repo=output_repo),
+        )
+    return {
+        'manager': os.path.join(repo_destination_path,
+                                'packages-urls/manager-packages.yaml'),
+        'agent': os.path.join(repo_destination_path,
+                              'packages-urls/agent-packages.yaml'),
+    }
+
+
+def download_file(source, destination):
+    """
+        Download a specific file.
+    """
+    print('Downloading {url}...'.format(url=source))
```

Comment:
i'd prefer to have the part under multiprocessing to not output anything; with print i think it's fairly safe, but if we ever transition to logging, we'll start having deadlock problems